parsl/app/futures.py

```diff
@@ -30,15 +38,23 @@ def parent_callback(self, parent_fu):
         Returns:
             - None
         """
-
         e = parent_fu._exception
         if e:
             self.set_exception(e)
         else:
             self.set_result(self.file_obj)
+            # only update the file object if it is a file
+            if self.data_flow_kernel.file_provenance and self.file_obj.scheme == 'file' and os.path.isfile(self.file_obj.filepath):
+                if not self.file_obj.timestamp:
+                    self.file_obj.timestamp = datetime.fromtimestamp(stat(self.file_obj.filepath).st_ctime, tz=timezone.utc)
+                if not self.file_obj.size:
+                    self.file_obj.size = stat(self.file_obj.filepath).st_size
+                if not self.file_obj.md5sum:
+                    self.file_obj.md5sum = md5(open(self.file_obj, 'rb').read()).hexdigest()
+            self.data_flow_kernel.register_as_output(self.file_obj, self.app_fut.task_record)
 
     @typeguard.typechecked
-    def __init__(self, fut: Future, file_obj: File, tid: int) -> None:
+    def __init__(self, fut: Future, file_obj: File, dfk: "DataFlowKernel", tid: Optional[int] = None, app_fut: Optional[Future] = None) -> None:
```

Comment:
Passing the DFK to the future smells off to me. Have you considered using the `add_done_callback` to set a method that interacts with the DFK?